File: 2197-replace-non-coprime-numbers-in-array/2197-replace-non-coprime-numbers-in-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var replaceNonCoprimes = function(nums) {
  const stack = [];
  
  for (let num of nums) {
    while (stack.length) {
      const g = gcd(num, stack.at(-1));
      if (g === 1) break;
      
      num *= stack.pop() / g;
    }
    
    stack.push(num);
  }
  
  return stack;
};

const gcd = (a, b) => !b ? a : gcd(b, a % b);

File: 2197-replace-non-coprime-numbers-in-array/2197-replace-non-coprime-numbers-in-array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        for num in nums:
            while stack:
                g = math.gcd(num, stack[-1])
                if g == 1:
                    break
                num *= stack.pop() // g
            stack.append(num)
        return stack